### ISSUE TITLE
Bug: async updates cause unwanted history

### ISSUE THREAD

**Issue by montlebalm** [facebook/lexical#4171]
There's a class of bugs involving async actions that leaves the history stack in an unwanted state. The problem occurs when a future update needs to replace a past update such as when uploading an image. Typically, an image node is inserted with a loading state because we don't know the `src` until it's uploaded. Eventually, we'll set the `src` as the result of an async action. If the async update is not merged then it's possible to "undo" back to the image's initial loading state which the user would never want.

Updates that use `discrete: true` and the `history-merge` tag will properly squash the update, but _only if there have been no other updates in between_. There's no way to merge a new update into a previous update of arbitrary position in the history stack.

The current way to solve this is to track data in a separate map keyed by node key. It's cumbersome and feels like something that the library should provide a built-in solution for.

Lexical version: 0.9.0 (and earlier)

## Steps To Reproduce

1. Insert an image node with a loading indicator. Asynchronously upload the desired image to your backend of choice.
2. Make random edits unrelated to the image
3. When the async upload finishes, update the placeholder image node with the new `src` attribute.
4. Trigger "undo" to reset the editor to state (2) where the image is still loading

Link to code example: https://codesandbox.io/s/lexical-plain-text-example-forked-usgq8e?file=/src/Editor.js

## The current behavior

The user can "undo" back to an unwanted loading state.

## The expected behavior

Previous history should be overwritten so that the user cannot return to unwanted loading states.

**Comment by thegreatercurve** [comment]
We allow uses to create their own history states with `createEmptyHistoryState`. You can then access it via context using something like the below which is on the playground: 
```ts
  const {historyState} = useSharedHistoryContext();
```
This will give you access to the undo and redo stacks, which you can probably manually manipulate to edit existing editor states.

**Comment by montlebalm** [comment]
> we allow uses to create their own history states with `createEmptyHistoryState` for these sort of edge cases

In my opinion, updating the editor as a result of an async action is not an edge case. It's anecdotal, but I've seen several people with similar issues in the Lexical Discord.

**Comment by kevinansfield** [comment]
@montlebalm did you get any further with handling this? I have similar questions around the handling of preview+loading progress states within nodes without causing a bunch of unwanted and problematic history states

**Comment by montlebalm** [comment]
Unfortunately no. I was hoping to get a response to my last comment regarding the expected way to handle these async updates.

Our current workaround is to keep this kind of data out of lexical as much as possible. For example, it feels natural to store an image's `src` on its node, but we have to keep it as React state since it's modified once the upload completes. It's inconvenient and makes serialization more difficult.

**Comment by acywatson** [comment]
I see that we didn't get a full resolution here, so I'm reopening so we can keep track of it

**Comment by acywatson** [comment]
>Our current workaround is to keep this kind of data out of lexical as much as possible. For example, it feels natural to store >an image's src on its node, but we have to keep it as React state since it's modified once the upload completes. It's >inconvenient and makes serialization more difficult.

I don't think this is bad. One strategy we make use of internally that bypasses this problem completely is storing some sort of ID on the node, then using that to fetch or generate a URL at runtime and encapsulating the logic for doing that in the React component. @fantactuka is working on documenting this pattern.

I realize that it suggests a change in architecture, and I'd love to get you a solution that doesn't force that, but I did want to call this one out. It's not without it's own complexities, though.

**Comment by acywatson** [comment]
Maybe a different history tag with slightly different semantics (like history-ignore or history-discard) could solve this

**Comment by hipstersmoothie** [comment]
Throwing in my use case:

1. I have special element nodes that get numbered 
2. these nodes only exist at the root
3. I use registerMutationListener to find when the nodes are added and deleted
4. I update the numbers

With no way to "merge" undo states all of my updates to the numbers create their own entry in the undo stack

**Comment by hipstersmoothie** [comment]
Oh history merge worked well for me. Should be documented

**Comment by fantactuka** [comment]
Not really sure how this can be built-in, given async action can have any delay you'd have to go through whole history undo/redo stacks, find that node and patch it. So it has to be some sort of `patch` callback that will alter all historical snapshots exposed through API? Instead I'd suggest to go with what @thegreatercurve suggested above and use `createEmptyHistoryState`:

1. Create own history stack so that you have pointer to it
2. When async action has finished you run editor.update with `historic` tag (a.k.a. history-ignore)
3. Run through history stack (since now you have pointer) and alter node value there

I'd also consider caching async request results for few reasons: you might copy-paste node while it's still running async action and you'd likely want to reuse existing running request instead of starting a new one.

**Comment by montlebalm** [comment]
I think rewriting history could be a fine solution provided there are examples. Even better if lexical exports a helper utility.

Here's what I was able to come up with based on the natural language description of the process:
```typescript
//
// Utility
//

const updateNode = <TNode extends LexicalNode>(
  stack: HistoryStateEntry,
  nodeKey: NodeKey,
  updater: (node: TNode) => void
): boolean => {
  //  Find the node in history. If we can't find it then we're past the point
  //  at which it was created.
  const node = stack.editorState.read(() => $getNodeByKey<TNode>(nodeKey));
  if (!node) return false;

  stack.editor.update(
    () => {
      updater(node);
      $addUpdateTag('history-ignore');
    },
    { discrete: true }
  );

  return true;
};

const updateHistoricalNode = <TNode extends LexicalNode>(
  history: HistoryState,
  nodeKey: NodeKey,
  updater: (node: TNode) => void
) => {
  // Update current state
  if (history.current) {
    updateNode(history.current, nodeKey, updater);
  }

  // Update undo history
  for (const stack of history.undoStack) {
    const updated = updateNode(stack, nodeKey, updater);
    if (!updated) break;
  }

  // Update redo history
  for (const stack of history.redoStack) {
    const updated = updateNode(stack, nodeKey, updater);
    if (!updated) break;
  }
};

//
// Usage
//

uploadImage(url).then((serverId) => {
  updateHistoricalNode(history, nodeKey, (node) => {
    node.setServerId(serverId);
  });
});
```

Does that look right? It could potentially cause _a lot_ of updates based on the size of the undo/redo stack. Any suggestions on how to make it more efficient?

**Comment by montlebalm** [comment]
> 2. When async action has finished you run editor.update with `historic` tag (a.k.a. history-ignore)

Should "history-ignore" be "history-merge"? I don't see the ignore tag in the codebase.

**Comment by OriginalEXE** [comment]
I've arrived here because I have a similar, but not same use case.

In the editor a user can type `/`, which shows a menu, from which they can select an action. Actions are async, so upon choosing the action I insert a temporary "loading indicator" node in place of `/`. When a server responds, I insert the response as a text node, replacing the "loading indicator" node. I would not want `ctrl + z` to get user back to a state where the "loading indicator" node is visible, it should either be a final state, or the state before they typed `/`.

**Comment by montlebalm** [comment]
@fantactuka @thegreatercurve does my example look like what you'd expect? Any suggestions for improvement?

**Comment by montlebalm** [comment]
@fantactuka @thegreatercurve just following-up to see if my proposed approach matches your expectations

**Comment by montlebalm** [comment]
@fantactuka @thegreatercurve checking in again to see if y'all have any guidance on how to do the proposed history rewrite
